### PR TITLE
Go: Fix data flow through variable defined in type switch guard

### PR DESCRIPTION
**Pull request by owen-mc**
Reported [here](https://github.com/github/codeql/issues/15350).

Consider a typical type switch statement.

```go
	switch inputWithType := input.(type) {
	case Type1:
		output1 = f(inputWithType)
	case Type2:
		output1 = g(inputWithType)
	}
```

It looks like a new variable `inputWithType` is defined in the guard, and is used in the case clauses. But the go compiler can't represent it like that internally because `inputWithType` has a different type in each case clause. So it has to make a variable for each case clause. The CodeQL Go library is not currently dealing with this. We end up with an object for `inputWithType` which doesn't have a declaration, which means that there is no `SsaDefinition` relating to it. This causes data flow to not work.

I tried fixing this in the extractor, but it was a bit messy, and I realised it was better to accept that these variables don't have an explicit declaration and add an `IR::Instruction` for their implicit declaration. This causes SSA to work properly again. I have added tests for control flow and data flow.

### CODE REVIEW

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll [hunk 1]
```diff
@@ -306,6 +306,16 @@ newtype TControlFlowNode =
...
    */
   MkCaseCheckNode(CaseClause cc, int i) { exists(cc.getExpr(i)) } or
+  /**
+   * A control-flow node that represents the implicit declaration of the variable for case clause
+   * `cc` of type switch statement `ts`, and its assignment of the value from the guard.
+   */
+  MkTypeSwitchImplicitVariable(CaseClause cc, Expr switchExpr) {
+    exists(TypeSwitchStmt ts | ts.getAssign() instanceof DefineStmt |
+      cc = ts.getACase() and
+      switchExpr = ts.getAssign().(DefineStmt).getRhs().(TypeAssertExpr).getExpr()
```
Comment (smowton) [inline go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:316]:
```suggestion
    exists(TypeSwitchStmt ts, DefineStmt ds | ds = ts.getAssign() |
      cc = ts.getACase() and
      switchExpr = ds.getRhs().(TypeAssertExpr).getExpr()
```

**Comment by smowton** [comment]
Liking that we apparently don't even need to touch the SSA layer -- this becomes just another instruction that defines a fresh variable 👍

File: go/extractor/extractor.go [hunk 2]
```diff
@@ -132,9 +132,7 @@ func ExtractWithFlags(buildFlags []string, patterns []string) error {
 	pkgsNotFound := make([]string, 0, len(pkgs))
 
 	// Do a post-order traversal and extract the package scope of each package
-	packages.Visit(pkgs, func(pkg *packages.Package) bool {
-		return true
-	}, func(pkg *packages.Package) {
+	packages.Visit(pkgs, nil, func(pkg *packages.Package) {
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/extractor/extractor.go:135]: Done in https://github.com/github/codeql/pull/16123

**Comment by owen-mc** [comment]
I found a way to identify the implicitly declared objects without a new table. This simplifies the PR a bit.

File: go/ql/lib/change-notes/2024-04-06-type-switch-implicitly-declared-variables.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Data flow through variables declared in statements of the form `x := y.(type)` at the beginning of type switches was not being followed. This has now been fixed, which may result in more alerts.
```
Comment (smowton) [inline go/ql/lib/change-notes/2024-04-06-type-switch-implicitly-declared-variables.md:4]:
```suggestion
* Data flow through variables declared in statements of the form `x := y.(type)` at the beginning of type switches has been fixed, which may result in more alerts.
```

**Review by smowton — APPROVED** [review]
QL changes look good; trivial suggestion re: change note. Haven't examined the test expectations in detail.

**Review by smowton — APPROVED** [review]

**Comment by owen-mc** [comment]
QA results all look good.

**Pull request merged**